File: backend/api/models/anomaly.py
"""
SatInsight AI — Anomaly Pydantic Models

Response schemas for anomaly detection and AI insight endpoints.
"""

from typing import Any, Optional
from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Individual anomaly row
# ---------------------------------------------------------------------------

class AnomalyRow(BaseModel):
    """One telemetry row annotated with anomaly scores and risk level."""
    row_index: int
    timestamp: Optional[str] = None
    temperature: Optional[float] = None
    radiation: Optional[float] = None
    pressure: Optional[float] = None
    battery_level: Optional[float] = None
    signal_strength: Optional[float] = None
    velocity: Optional[float] = None
    altitude: Optional[float] = None
    if_score: float
    max_zscore: float
    risk_level: str                 # "Normal" | "Warning" | "Critical"
    anomaly_params: list[str]       # parameters that breached Z-score threshold


# ---------------------------------------------------------------------------
# Analyze response
# ---------------------------------------------------------------------------

class RiskCounts(BaseModel):
    Normal: int
    Warning: int
    Critical: int
    total_anomalies: int
    total_rows: int


class AnalyzeResponse(BaseModel):
    session_id: str
    dataset_name: str
    row_count: int
    risk_counts: RiskCounts
    param_anomaly_counts: dict[str, int]
    message: str


# ---------------------------------------------------------------------------
# Anomalies list (paginated)
# ---------------------------------------------------------------------------

class AnomalyListResponse(BaseModel):
    session_id: str
    page: int
    page_size: int
    total_anomalies: int
    total_pages: int
    filter_risk: Optional[str]      # None = all, "Warning", "Critical"
    anomalies: list[dict[str, Any]]


# ---------------------------------------------------------------------------
# Insight card
# ---------------------------------------------------------------------------

class InsightCard(BaseModel):
    id: str
    title: str
    description: str
    severity: str                   # "info" | "warning" | "critical"
    parameter: str                  # primary affected parameter or "multi"
    value_range: str
    recommended_action: str
    row_start: int
    row_end: int
    event_rows: int


class InsightsResponse(BaseModel):
    session_id: str
    dataset_name: str
    total_insights: int
    insights: list[InsightCard]
    analysis_run: bool              # False if analyze hasn't been called yet

File: frontend/src/types/anomaly.ts
// SatInsight AI — TypeScript Types: Anomaly & Insights

export type RiskLevel = 'Normal' | 'Warning' | 'Critical'
export type Severity = 'info' | 'warning' | 'critical'

export interface RiskCounts {
  Normal: number
  Warning: number
  Critical: number
  total_anomalies: number
  total_rows: number
}

export interface AnalyzeResponse {
  session_id: string
  dataset_name: string
  row_count: number
  risk_counts: RiskCounts
  param_anomaly_counts: Record<string, number>
  message: string
}

export interface AnomalyRow {
  row_index: number
  timestamp: string | null
  temperature: number | null
  radiation: number | null
  pressure: number | null
  battery_level: number | null
  signal_strength: number | null
  velocity: number | null
  altitude: number | null
  if_score: number
  max_zscore: number
  risk_level: RiskLevel
  anomaly_params: string[]
  [key: string]: string | number | string[] | null
}

export interface AnomalyListResponse {
  session_id: string
  page: number
  page_size: number
  total_anomalies: number
  total_pages: number
  filter_risk: string | null
  anomalies: AnomalyRow[]
}

export interface InsightCard {
  id: string
  title: string
  description: string
  severity: Severity
  parameter: string
  value_range: string
  recommended_action: string
  row_start: number
  row_end: number
  event_rows: number
}

export interface InsightsResponse {
  session_id: string
  dataset_name: string
  total_insights: number
  insights: InsightCard[]
  analysis_run: boolean
}

export interface AnomalySummary {
  session_id: string
  risk_counts: RiskCounts
  param_anomaly_counts: Record<string, number>
}
